src/corehost/corehost.cpp

```diff
@@ -65,44 +66,90 @@ bool is_exe_enabled_for_execution(const pal::string_t& own_path)
         return false;
     }
 
-    pal::string_t own_name = get_filename(own_path);
-    pal::string_t own_dll_filename = get_executable(own_name) + _X(".dll");
+    trace::info(_X("The managed DLL bound to this executable is: '%s'"), app_dll->c_str());
+    return true;
+}
 
-    if (pal::strcasecmp(own_dll_filename.c_str(), pal_binding.c_str()) != 0)
+bool resolve_app_root(const pal::string_t& host_path, pal::string_t* out_app_root, bool* requires_v2_hostfxr_interface)
+{
+    // For self-contained, the startupconfig.json specifies app_root which is used later to assign dotnet_root
+    // For framework-dependent, the startupconfig.json specifies app_root which does not affect dotnet_root
+    pal::string_t config_path = strip_executable_ext(host_path);
+    config_path += _X(".startupconfig.json");
+    startup_config_t startup_config;
+    startup_config.parse(config_path);
+    if (!startup_config.is_valid())
     {
-        trace::error(_X("The managed DLL bound to this executable: '%s', did not match own name '%s'."), pal_binding.c_str(), own_dll_filename.c_str());
         return false;
     }
 
-    trace::info(_X("The managed DLL bound to this executable is: '%s'"), pal_binding.c_str());
+    if (startup_config.get_app_root().empty())
+    {
+        out_app_root->assign(get_directory(host_path));
+    }
+    else
+    {
+        *requires_v2_hostfxr_interface = true;
+        if (pal::is_path_rooted(startup_config.get_app_root()))
+        {
+            out_app_root->assign(startup_config.get_app_root());
+        }
+        else
+        {
+            out_app_root->assign(get_directory(host_path));
+            append_path(out_app_root, startup_config.get_app_root().c_str());
+        }
+        if (!pal::realpath(out_app_root))
+        {
+            trace::error(_X("The app root [%s] specified in [%s] does not exist."),
+                out_app_root->c_str(), config_path.c_str());
+            return false;
+        }
+    }
+
     return true;
 }
-#endif // FEATURE_APPHOST
+#endif
 
-pal::string_t resolve_fxr_path(const pal::string_t& own_dir)
+bool resolve_fxr_path(const pal::string_t& host_path, const pal::string_t& app_root, pal::string_t* out_dotnet_root, pal::string_t* out_fxr_path)
 {
+    pal::string_t host_dir;
+    host_dir.assign(get_directory(host_path));
+
 #if FEATURE_APPHOST
-    pal::string_t fxr_path;
-    if (library_exists_in_dir(own_dir, LIBFXR_NAME, &fxr_path))
+    // If a hostfxr exists in app_root, then assumed self-contained.
+    if (library_exists_in_dir(app_root, LIBFXR_NAME, out_fxr_path))
     {
-        trace::info(_X("Resolved fxr [%s]..."), fxr_path.c_str());
-        return fxr_path;
+        trace::info(_X("Resolved fxr [%s]..."), out_fxr_path->c_str());
+        out_dotnet_root->assign(app_root);
+        return true;
     }
 
-    trace::error(_X("A fatal error occurred, the required library %s could not be found at [%s]"), LIBFXR_NAME, own_dir.c_str());
```

Comment:
Do we still write an error in this case?

Reply:
The error would be alone the lines of "A fatal error occurred, the folder [%s] does not exist", which is not useful depending on the context. Adding a more detailed error.